docs/architecture/adr-050-sign-mode-textual.md

```diff
@@ -129,6 +130,44 @@ Msg (<int>/<int>): <string>           // E.g. Msg (1/2): bank v1beta1 send coins
 End of transaction messages
 ```
 
+### 9. Custom `Msg` Renderers
+
+Application developers may choose to not follow default renderer value output for their own `Msg`s. In this case, they can implement their own custom `Msg` renderer. This is similar to [EIP4430](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-4430.md), where the smart contract developer chooses the description string to be shown to the end user.
+
+This is done by setting the `cosmos.msg.v1.textual.custom_renderer` Protobuf option to a non-empty string. This option CAN ONLY be set on a Protobuf message representing transaction message object (implementing `sdk.Msg` interface).
+
+```proto
+message MsgFooBar {
+  // Optional comments to describe in human-readable language the formatting
+  // rules of the custom renderer.
+  option (cosmos.msg.v1.textual.custom_renderer) = "<unique algorithm identifier>";
+
+  // proto fields
+}
+```
+
+When this option is set on a `Msg`, a registered function will transform the `Msg` into an array of one or more strings, which MAY use the key/value format (described in point #3) with the expert field prefix (described in point #5). These strings MAY be rendered from a `Msg` field using a default value renderer, or they may be generated from several fields using custom logic.
```

Comment:
```suggestion
When this option is set on a `Msg`, a registered function will transform the `Msg` into an array of one or more strings, which MUST use the key/value format (described in point #3) and MAY use the expert field prefix (described in point #5). These strings MAY be rendered from a `Msg` field using a default value renderer, or they may be generated from several fields using custom logic.
```

any opinions on MAY v MUST here for the `key: value` format? If we do MUST, it'll probably make client UI lives easier.

Reply:
+1 for MUST

Reply:
Depends on what guarantees you want to make for the signing agent. I think the freedoms of a custom renderer should match the freedoms given the default renderer - the signing agent needs to handle either, and we can make things less complicated for it by not having a difference between the range of formats emitted by a custom vs default renderer.

Reply:
After thinking about this more, the custom renderer must have the ability to generate multi-screen output, including aggregate start and end screens, so not everything will be key/value.